overlord/ifacestate/handlers.go

```diff
@@ -1258,6 +1258,46 @@ func (m *InterfaceManager) doGadgetConnect(task *state.Task, _ *tomb.Tomb) error
 	return nil
 }
 
+// doHotplugUpdateSlot updates static attributes of a hotplug slot for given device.
+func (m *InterfaceManager) doHotplugUpdateSlot(task *state.Task, _ *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	ifaceName, hotplugKey, err := getHotplugAttrs(task)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get hotplug task attributes: %s", err)
+	}
+
+	var attrs map[string]interface{}
+	if err := task.Get("slot-attrs", &attrs); err != nil {
+		return fmt.Errorf("internal error: cannot get slot-attrs attribute for device %s, interface %s: %s", hotplugKey, ifaceName, err)
+	}
+
+	stateSlots, err := getHotplugSlots(st)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain hotplug slots: %v", err)
+	}
+
+	slot, err := m.repo.UpdateStaticSlotAttrs(ifaceName, hotplugKey, attrs)
+	if err != nil {
+		return err
+	}
+	if slot == nil {
+		return nil
+	}
+
+	if slotSpec, ok := stateSlots[slot.Name]; ok {
```

Comment:
I'm a bit surprised by the order here, I would have expected this check to happen before we consider the repo.  Is the state or the repo authoritative here?  the repo is runtime only...  We don't have yet the code that adds a slot to state,  is that right?  (I suppose that relates also to the discussion around implicit slots vs hotplug slots we had last week) I don't see code to add a slot in hotplug-connect, does it need to be there? what is the plan about that? or have a hotplug-add-slot task?

Reply:
The order doesn't matter because the basic invariant is that repo always reflects `hotplug-slots` in the state (stateSlots). Interface manager startup code restores all last known hotplug slots from `hotplug-slots` state and they are reflected in the repository; this happens in ifacemgr initialize / addSnaps in the last `hotplug` PR. Enumeration affects this asynchronously by creating tasks add/remove/update slots as needed.

The adding of the slot to hotplug-slots in the state and repo is also implemented in the `hotplug` PR, but after your comment I realized I've a bug there in how it's happening with relation to hotplug tasks: if the slot doesn't exist in hotplug-state I create it immediately in state and repo (outside of hotplug tasks); also the decision whether slot should be updated via this `hotplug-update-slot` task is made via comparision of attributes made immediately as the device is detected. This is wrong as it bypasses the queing and will lead incorrect operations on edge cases (e.g. inserting/removing/inserting a device very quickly). I need to refactor that with a new task type.

Reply:
Actually, scratch the first part of my comment. I was inaccurate with that explanation, and also have an error around that in the implementation in the other hotplug PR.

The hotplug update semantics (as implemented atm) require that if we detect a change in slot attributes when device is plugged but was never unplugged (i.e. snapd is restarted and we detect a change when enumerating) then we need to recreate old slot and old connections in the repo just to run disconnects and hooks on them in a regular way. The sequence of tasks in such case will be: hotplug-add-slot -> hotplug-disconnect -> hotplug-update-slot -> hotplug-connect.